src/google/protobuf/compiler/command_line_interface_unittest.cc

```diff
@@ -940,6 +959,53 @@ TEST_F(CommandLineInterfaceTest, WriteTransitiveDescriptorSetWithSourceInfo) {
   EXPECT_TRUE(descriptor_set.file(1).has_source_code_info());
 }
 
+TEST_F(CommandLineInterfaceTest, WriteDependencyManifestFile) {
+  CreateTempFile("foo.proto",
+    "syntax = \"proto2\";\n"
+    "message Foo {}\n");
+  CreateTempFile("bar.proto",
+    "syntax = \"proto2\";\n"
+    "import \"foo.proto\";\n"
+    "message Bar {\n"
+    "  optional Foo foo = 1;\n"
+    "}\n");
+
+  Run("protocol_compiler --dependency_out=$tmpdir/manifest "
+      "--test_out=$tmpdir --proto_path=$tmpdir bar.proto");
+
+  ExpectNoErrors();
+
+  ExpectFileContent("manifest",
+                    "$tmpdir/manifest: $tmpdir/foo.proto\\\n"
```

Comment:
On Wed, Feb 4, 2015 at 2:33 PM, dmaclach <notifications@github.com> wrote:

> In src/google/protobuf/compiler/command_line_interface_unittest.cc
> <https://github.com/google/protobuf/pull/193#discussion_r24126637>:
>
> > +    "syntax = \"proto2\";\n"
> > +    "message Foo {}\n");
> > +  CreateTempFile("bar.proto",
> > +    "syntax = \"proto2\";\n"
> > +    "import \"foo.proto\";\n"
> > +    "message Bar {\n"
> > +    "  optional Foo foo = 1;\n"
> > +    "}\n");
> > +
> > +  Run("protocol_compiler --dependency_out=$tmpdir/manifest "
> > +      "--test_out=$tmpdir --proto_path=$tmpdir bar.proto");
> > +
> > +  ExpectNoErrors();
> > +
> > +  ExpectFileContent("manifest",
> > +                    "$tmpdir/manifest: $tmpdir/foo.proto\\\n"
>
> normally the file would be:
> ... : \
> \
> \
> ...
>
> where for a main.c of:
> #include "main.h"
>
> int main() {
> return 0;
> }
>
> cc -MMD main.c
>
> you are going to get:
>
> main.o: main.c main.h
>
> our code is not going to set correctly. I would expect for the cpp
> compiler that we would actually generate
>
> foo.pb.cc foo.pb.h : foo.proto
>
> Thoughts? I think the current code will actually work for my needs, but it
> is certainly not a correct solution as far as I can tell.
>
Protoc has one difference from gcc: protoc can generate different set of
files for different languages, where as gcc is tied to C/C++. This means
the output of protoc varies from language. And sometimes it's may not be
tied with any language at all. For example, users may just want to find all
imports of a certain .proto file and build a protodb for them. For this
reason I think it's better to avoid any language specific files in the
generated dependency file and make it a general purpose dependency file
that could be used for any language.


I'm working from "
> http://make.mad-scientist.net/papers/advanced-auto-dependency-generation/
> ".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/protobuf/pull/193/files#r24126637>.
>